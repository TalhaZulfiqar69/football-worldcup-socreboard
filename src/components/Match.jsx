import React, { useState } from "react";
import {
  Button,
  Input,
  FormControl,
  FormLabel,
  Heading,
  Box,
  Grid,
  GridItem,
  Flex,
  Card,
  CardHeader,
  CardBody,
  Stack,
  StackDivider,
} from "@chakra-ui/react";

import MatchSummary from "./MatchSummary";

const Match = ({ match, onUpdateScore, onFinish, matches }) => {
  const [homeScore, setHomeScore] = useState(0);
  const [homeTeam, setHomeTeam] = useState("");
  const [awayScore, setAwayScore] = useState(0);
  const [awayTeam, setAwayTeam] = useState("");
  const [showSummaries, setShowSummaries] = useState(false);

  const handleUpdateScore = () => {
    onUpdateScore(homeScore, awayScore, homeTeam, awayTeam);
  };

  const handleToggleSummaries = () => {
    setShowSummaries(!showSummaries);
  };
  return (
    <Box>
      <Card mb={4}>
        <CardHeader>
          <Heading as="h4" size="md" textAlign={"center"}>
            {match.homeTeam} VS {match.awayTeam}
          </Heading>
        </CardHeader>

        <CardBody>
          <Stack divider={<StackDivider />} spacing="4">
            <Heading as={"h5"} size={"md"}>
              {match.homeTeam} {homeScore} - {match.awayTeam} {awayScore}
            </Heading>

            <Grid templateColumns="repeat(2, 1fr)" gap={6}>
              <GridItem w="100%">
                <FormControl>
                  <FormLabel fontWeight={"bold"}>Home Team Name:</FormLabel>
                  <Input
                    placeholder="Enter home team name"
                    type="text"
                    value={homeTeam}
                    onChange={(e) => setHomeTeam(e.target.value)}
                  />
                </FormControl>
              </GridItem>
              <GridItem w="100%">
                <FormControl>
                  <FormLabel fontWeight={"bold"}>Home Team Score:</FormLabel>
                  <Input
                    type="number"
                    value={homeScore}
                    onChange={(e) => setHomeScore(e.target.value)}
                  />
                </FormControl>
              </GridItem>
            </Grid>

            <Grid templateColumns="repeat(2, 1fr)" gap={6}>
              <GridItem w="100%">
                <FormControl>
                  <FormLabel fontWeight={"bold"}>Away Team Name:</FormLabel>
                  <Input
                    placeholder="Enter away team name"
                    type="text"
                    value={awayTeam}
                    onChange={(e) => setAwayTeam(e.target.value)}
                  />
                </FormControl>
              </GridItem>
              <GridItem w="100%">
                <FormControl>
                  <FormLabel fontWeight={"bold"}>Away Team Score:</FormLabel>
                  <Input
                    type="number"
                    value={awayScore}
                    onChange={(e) => setAwayScore(e.target.value)}
                  />
                </FormControl>
              </GridItem>
            </Grid>

            <Flex mt={4} mb={4}>
              <Button
                onClick={handleUpdateScore}
                mr={14}
                bgGradient="linear(to-l, #7928CA, #FF0080)"
                _hover={{
                  bgGradient: "linear(to-l, #7928CA, #FF0080)",
                }}
                color={"#ffffff"}
              >
                Update Score
              </Button>
              <Button
                onClick={onFinish}
                mr={14}
                bgGradient="linear(to-l, #7928CA, #FF0080)"
                _hover={{
                  bgGradient: "linear(to-l, #7928CA, #FF0080)",
                }}
                color={"#ffffff"}
              >
                Finish Match
              </Button>
              {matches?.length ? (
                <Button
                  onClick={handleToggleSummaries}
                  mb={4}
                  bgGradient="linear(to-l, #7928CA, #FF0080)"
                  _hover={{
                    bgGradient: "linear(to-l, #7928CA, #FF0080)",
                  }}
                  color={"#ffffff"}
                >
                  View Match Summaries
                </Button>
              ) : (
                ""
              )}
            </Flex>
          </Stack>
        </CardBody>
      </Card>

      {showSummaries && <MatchSummary matches={matches} />}
    </Box>
  );
};

export default Match;
