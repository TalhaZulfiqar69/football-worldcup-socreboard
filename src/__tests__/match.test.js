import { render, screen, fireEvent } from "@testing-library/react";
import { axe } from "jest-axe";
import "jest-axe/extend-expect";
import { ChakraProvider, extendTheme } from "@chakra-ui/react";
import Match from "../components/Match";
import mockTheme from "../chakra-ui-mock-theme";

const extendedTheme = extendTheme({ ...mockTheme });
// TESTCASE # 1: Check initial state rendering
// DESCRIPTION: Here I am testing if the component renders withour errors and the the initial state is set correctly or not.
test("renders component with initial state", () => {
  render(
    <ChakraProvider theme={extendedTheme}>
      <Match
        match={{}}
        onUpdateScore={() => {}}
        onFinish={() => {}}
        matches={[]}
      />
    </ChakraProvider>
  );

  // Assertions for home team input
  const homeTeamInput = screen.getByPlaceholderText("Enter home team name");
  expect(homeTeamInput).toBeInTheDocument();
  expect(homeTeamInput.value).toBe("");

  const homeScoreInput = screen.getByLabelText("Home Team Score:");
  expect(homeScoreInput).toBeInTheDocument();
  expect(homeScoreInput.value).toBe("0");

  // Assertions for away team input
  const awayTeamInput = screen.getByPlaceholderText("Enter away team name");
  expect(awayTeamInput).toBeInTheDocument();
  expect(awayTeamInput.value).toBe("");

  const awayScoreInput = screen.getByLabelText("Away Team Score:");
  expect(awayScoreInput).toBeInTheDocument();
  expect(awayScoreInput.value).toBe("0");
});

// TESTCASE # 2: Input and State Interaction
// DESCRIPTION: Here I am testing that input fields interact with state correctly or not.
test("input fields interact with state", () => {
  render(
    <ChakraProvider theme={extendedTheme}>
      <Match
        match={{}}
        onUpdateScore={() => {}}
        onFinish={() => {}}
        matches={[]}
      />
    </ChakraProvider>
  );

  // Simulate user input
  const homeTeamInput = screen.getByPlaceholderText("Enter home team name");
  fireEvent.change(homeTeamInput, { target: { value: "Home Team" } });
  expect(homeTeamInput.value).toBe("Home Team");

  const homeScoreInput = screen.getByLabelText("Home Team Score:");
  fireEvent.change(homeScoreInput, { target: { value: "2" } });
  expect(homeScoreInput.value).toBe("2");

  const awayTeamInput = screen.getByPlaceholderText("Enter away team name");
  fireEvent.change(awayTeamInput, { target: { value: "Away Team" } });
  expect(awayTeamInput.value).toBe("Away Team");

  const awayScoreInput = screen.getByLabelText("Away Team Score:");
  fireEvent.change(awayScoreInput, { target: { value: "2" } });
  expect(awayScoreInput.value).toBe("2");
});

// TESTCASE # 3: Button clicks
// DESCRIPTION: Here I am testing that button clicks trigger the expected functions or not.
test("button clicks trigger functions", () => {
  const mockUpdateScore = jest.fn();
  const mockOnFinish = jest.fn();
  const mockMatches = [];

  render(
    <ChakraProvider theme={extendedTheme}>
      <Match
        match={{}}
        onUpdateScore={mockUpdateScore}
        onFinish={mockOnFinish}
        matches={mockMatches}
      />
    </ChakraProvider>
  );

  // Simulate button clicks
  const updateScoreButton = screen.getByText("Update Score");
  fireEvent.click(updateScoreButton);
  expect(mockUpdateScore).toHaveBeenCalledTimes(1);

  const finishMatchButton = screen.getByText("Finish Match");
  fireEvent.click(finishMatchButton);
  expect(mockOnFinish).toHaveBeenCalledTimes(1);
});

// TESTCASE # 4: Toggle summaries
// DESCRIPTION: Here I am testing that when I click the button the match summaries are toggled correctly or not.
test("toggle summaries button works", () => {
  const mockMatches = [
    { homeScore: 1, awayScore: 2 },
    { homeScore: 3, awayScore: 4 },
  ];

  const { rerender } = render(
    <ChakraProvider theme={extendedTheme}>
      <Match
        match={{}}
        onUpdateScore={() => {}}
        onFinish={() => {}}
        matches={mockMatches}
      />
    </ChakraProvider>
  );

  // Initial state, summaries should be hidden
  const summariesComponent = screen.queryByTestId("match-summary");
  expect(summariesComponent).toBeNull();

  // Click on the toggle summaries button
  const toggleSummariesButton = screen.getByText("View Match Summaries");
  fireEvent.click(toggleSummariesButton);

  rerender(
    <ChakraProvider theme={extendedTheme}>
      <Match
        match={{}}
        onUpdateScore={() => {}}
        onFinish={() => {}}
        matches={mockMatches}
      />
    </ChakraProvider>
  );

  const visibleSummariesComponent = screen.getByTestId("match-summary");
  expect(visibleSummariesComponent).toBeInTheDocument();
});

// TESTCASE # 5: Performance test
// DESCRIPTION: Here I am testing the component's performance by rendering a large number of components and measuring the time it takes.
test("renders multiple components without performance issues", () => {
  const mockMatches = new Array(1000).fill({ homeScore: 0, awayScore: 0 });

  const startTime = performance.now();
  render(
    <ChakraProvider theme={extendedTheme}>
      <Match
        match={{}}
        onUpdateScore={() => {}}
        onFinish={() => {}}
        matches={mockMatches}
      />
    </ChakraProvider>
  );
  const endTime = performance.now();

  const renderTime = endTime - startTime;
  expect(renderTime).toBeLessThan(100);
});

// TESTCASE # 6: Accessibility test
// DESCRIPTION:  Here I am testing the component for accessibility issues using tools like Jest Axe or React Testing Library's axe function.
test("component has no accessibility violations", async () => {
  const { container } = render(
    <ChakraProvider theme={extendedTheme}>
      <Match
        match={{}}
        onUpdateScore={() => {}}
        onFinish={() => {}}
        matches={[]}
      />
    </ChakraProvider>
  );
  const results = await axe(container);

  expect(results).toHaveNoViolations();
});

// TESTCASE # 6: Snapshot matching
// DESCRIPTION:  Here I am testing that if something change in component each time when testcases executes.
test("component snapshot matches", () => {
  const mockMatch = {
    homeTeam: "Team A",
    awayTeam: "Team B",
  };

  const { container } = render(
    <ChakraProvider theme={extendedTheme}>
      <Match
        match={mockMatch}
        onUpdateScore={() => {}}
        onFinish={() => {}}
        matches={[]}
      />
    </ChakraProvider>
  );

  expect(container).toMatchSnapshot();
});
