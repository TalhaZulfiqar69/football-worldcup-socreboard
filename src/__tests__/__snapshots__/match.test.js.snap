// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component snapshot matches 1`] = `
<div>
  <div
    class="css-0"
  >
    <div
      class="chakra-card css-169l99m"
    >
      <div
        class="chakra-card__header css-1sl53ol"
      >
        <h4
          class="chakra-heading css-1886hlo"
        >
          Team A
           VS 
          Team B
        </h4>
      </div>
      <div
        class="chakra-card__body css-1idwstw"
      >
        <div
          class="chakra-stack css-j7qwjs"
        >
          <h5
            class="chakra-heading css-18j379d"
          >
            Team A
             
            0
             - 
            Team B
             
            0
          </h5>
          <div
            class="chakra-stack__divider css-1q5cbl"
          />
          <div
            class="css-wclpxa"
          >
            <div
              class="css-8atqhb"
            >
              <div
                class="chakra-form-control css-1kxonj9"
                role="group"
              >
                <label
                  class="chakra-form__label css-1k48i1l"
                  for="field-:ro:"
                  id="field-:ro:-label"
                >
                  Home Team Name:
                </label>
                <input
                  class="chakra-input css-1cjy4zv"
                  id="field-:ro:"
                  placeholder="Enter home team name"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="css-8atqhb"
            >
              <div
                class="chakra-form-control css-1kxonj9"
                role="group"
              >
                <label
                  class="chakra-form__label css-1k48i1l"
                  for="field-:rp:"
                  id="field-:rp:-label"
                >
                  Home Team Score:
                </label>
                <input
                  class="chakra-input css-1cjy4zv"
                  id="field-:rp:"
                  type="number"
                  value="0"
                />
              </div>
            </div>
          </div>
          <div
            class="chakra-stack__divider css-1q5cbl"
          />
          <div
            class="css-wclpxa"
          >
            <div
              class="css-8atqhb"
            >
              <div
                class="chakra-form-control css-1kxonj9"
                role="group"
              >
                <label
                  class="chakra-form__label css-1k48i1l"
                  for="field-:rq:"
                  id="field-:rq:-label"
                >
                  Away Team Name:
                </label>
                <input
                  class="chakra-input css-1cjy4zv"
                  id="field-:rq:"
                  placeholder="Enter away team name"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="css-8atqhb"
            >
              <div
                class="chakra-form-control css-1kxonj9"
                role="group"
              >
                <label
                  class="chakra-form__label css-1k48i1l"
                  for="field-:rr:"
                  id="field-:rr:-label"
                >
                  Away Team Score:
                </label>
                <input
                  class="chakra-input css-1cjy4zv"
                  id="field-:rr:"
                  type="number"
                  value="0"
                />
              </div>
            </div>
          </div>
          <div
            class="chakra-stack__divider css-1q5cbl"
          />
          <div
            class="css-1m6s23t"
          >
            <button
              class="chakra-button css-y7t0mp"
              type="button"
            >
              Update Score
            </button>
            <button
              class="chakra-button css-y7t0mp"
              type="button"
            >
              Finish Match
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
